Holtron.Net/Network/Encryption/NetEncryptionAESGCM.cs

```diff
@@ -0,0 +1,82 @@
+using System.Security.Cryptography;
+
+namespace Holtron.Net.Network.Encryption
+{
+    public class NetEncryptionAESGCM : INetEncryption
+    {
+        // This implementation is from
+        // https://stackoverflow.com/questions/60889345/using-the-aesgcm-class
+        private readonly AesGcm aesGcm;
+
+        public NetEncryptionAESGCM(byte[] key)
+        {
+            var derivedKey = new Rfc2898DeriveBytes(key, new byte[8], 1000).GetBytes(16);
+            aesGcm = new AesGcm(derivedKey);
+        }
+
+        public NetEncryptionAESGCM(string key)
+        {
+            var derivedKey = new Rfc2898DeriveBytes(key, new byte[8], 1000).GetBytes(16);
+            aesGcm = new AesGcm(derivedKey);
+        }
+
+        public NetEncryptionAESGCM(Span<byte> key)
+        {
+            var derivedKey = new Rfc2898DeriveBytes(key.ToArray(), new byte[8], 1000).GetBytes(16);
+            aesGcm = new AesGcm(derivedKey);
+        }
+
+        public bool Decrypt(NetIncomingMessage message)
+        {
+            var unencryptedLengthBits = (int)message.ReadUInt32();
+            var encryptedDataLength = message.ReadInt32();
+            var encryptedData = message.ReadBytes(encryptedDataLength).AsSpan();
+            var nonce = encryptedData.Slice(0, 12);
+            var tag = encryptedData.Slice(12, 12);
+            var cipherText = encryptedData.Slice(24, encryptedData.Length - 24);
+            var plainTextMessage = new byte[cipherText.Length];
+
+            aesGcm.Decrypt(nonce, cipherText, tag, plainTextMessage);
+
+            message.Data = plainTextMessage.ToArray();
+            message.m_data = plainTextMessage.ToArray();
+            message.m_bitLength = unencryptedLengthBits;
+            message.m_readPosition = 0;
+
+            return true;
+        }
+
+        public void Encrypt(NetOutgoingMessage message)
+        {
+            var bytesToEncrypt = new byte[message.Data.Length];
+            message.Data.CopyTo(bytesToEncrypt, 0);
+            var unencryptedLengthBits = message.LengthBits;
+
+            var encryptedDataLength = 24 + bytesToEncrypt.Length;
+
+            Span<byte> encryptedData = encryptedDataLength < 1024
+                ? stackalloc byte[encryptedDataLength]
+                : new byte[encryptedDataLength].AsSpan();
+
+            var nonce = encryptedData.Slice(0, 12);
```

Comment:
Is the size of the `nonce` and `tag` always guaranteed to be 12 bytes? This might be variable depending on where the code is running.

Reply:
The nonce and tag are somethin' you set as part of the implementation. In the case of this, I'm defaulting to the smallest size allowable per the AesGcm class constants. I am gonna go ahead and move these to constants though so we can change them more quickly if needed.